File: LeetCode-Cpp/SnakesandLadders.cpp
class Solution {
public:
    int snakesAndLadders(vector<vector<int>>& board) {
        int n = board.size();
        queue<pair<int, int>> q;
        q.push({1, 0});
        vector<bool> visited(n * n + 1, false);
        visited[1] = true;

        while (!q.empty()) {
            auto [square, moves] = q.front();
            q.pop();

            if (square == n * n) return moves;

            for (int i = 1; i <= 6 && square + i <= n * n; ++i) {
                int next = square + i;
                auto [row, col] = getPosition(next, n);
                if (board[row][col] != -1) {
                    next = board[row][col];
                }

                if (!visited[next]) {
                    visited[next] = true;
                    q.push({next, moves + 1});
                }
            }
        }

        return -1;
    }

private:
    pair<int, int> getPosition(int s, int n) {
        int quot = (s - 1) / n;
        int rem = (s - 1) % n;
        int row = n - 1 - quot;
        int col = (quot % 2 == 0) ? rem : n - 1 - rem;
        return {row, col};
}
};

File: LeetCode-Python/SnakesandLadders.py
class Solution:
    def snakesAndLadders(self, board: List[List[int]]) -> int:
        n = len(board)

        def get_position(s):
            quot, rem = divmod(s - 1, n)
            row = n - 1 - quot
            col = rem if quot % 2 == 0 else n - 1 - rem
            return row, col

        visited = [False] * (n * n + 1)
        queue = deque()
        queue.append((1, 0))  # (current square, move count)
        visited[1] = True

        while queue:
            square, moves = queue.popleft()

            if square == n * n:
                return moves

            for i in range(1, 7):
                next_square = square + i
                if next_square > n * n:
                    continue

                r, c = get_position(next_square)

                # Go to destination of snake or ladder
                if board[r][c] != -1:
                    next_square = board[r][c]

                if not visited[next_square]:
                    visited[next_square] = True
                    queue.append((next_square, moves + 1))

        return -1  # Not reachable
